crates/ide/src/move_item.rs

```diff
@@ -0,0 +1,372 @@
+use std::iter::once;
+
+use hir::Semantics;
+use ide_db::{base_db::FileRange, RootDatabase};
+use syntax::{algo, AstNode, NodeOrToken, SyntaxElement, SyntaxKind, SyntaxNode};
+use text_edit::{TextEdit, TextEditBuilder};
+
+pub enum Direction {
+    Up,
+    Down,
+}
+
+// Feature: Move Item
+//
+// Move item under cursor or selection up and down.
+//
+// |===
+// | Editor  | Action Name
+//
+// | VS Code | **Rust Analyzer: Move item up**
+// | VS Code | **Rust Analyzer: Move item down**
+// |===
+pub(crate) fn move_item(
+    db: &RootDatabase,
+    range: FileRange,
+    direction: Direction,
+) -> Option<TextEdit> {
+    let sema = Semantics::new(db);
+    let file = sema.parse(range.file_id);
+
+    let item = file.syntax().covering_element(range.range);
+    find_ancestors(item, direction)
+}
+
+fn find_ancestors(item: SyntaxElement, direction: Direction) -> Option<TextEdit> {
+    let root = match item {
+        NodeOrToken::Node(node) => node,
+        NodeOrToken::Token(token) => token.parent(),
+    };
+
+    let ancestor = once(root.clone())
+        .chain(root.ancestors())
+        .filter_map(|ancestor| kind_priority(ancestor.kind()).map(|priority| (priority, ancestor)))
+        .max_by_key(|(priority, _)| *priority)
+        .map(|(_, ancestor)| ancestor)?;
+
+    move_in_direction(&ancestor, direction)
+}
+
+fn kind_priority(kind: SyntaxKind) -> Option<i32> {
```

Comment:
Those priorities are needed to, for example, move match arm instead of a statement, when cursor is inside an arm. Not sure if current implementation is good enough though

Reply:
Nope, it's bad. It worked on match arms good, yet everything besides match arms feels uncomfortable. Removed for now